### PR TITLE
Queue spool.

### PR DESCRIPTION
**Pull request by makasim**
PR adds an ability to use any queue-interop compatible MQ transport as an email spool. 

Here's use case example, the source code is [here](https://github.com/php-enqueue/enqueue-sandbox/tree/master/swiftmailer):

```php
<?php

// send.php
// could be run as GUSER=yourAccountName@gmail.com GPASS="yourGmailPassword" php send.php

require_once __DIR__.'/vendor/autoload.php';

$transport = new Swift_SpoolTransport(new \Swift_QueueSpool(
    (new \Enqueue\Fs\FsConnectionFactory('file://'.__DIR__.'/queue'))->createContext()
));

$mailer = new Swift_Mailer($transport);

$message = (new Swift_Message('Wonderful Subject'))
    ->setFrom(getenv('GUSER'))
    ->setTo(getenv('GUSER'))
    ->setBody('Here is the message itself. Sent at: '.date('Y-m-d H:i:s'))
;

$result = $mailer->send($message);
```

The message was queued, to send message for real we have to flush queue.

```php
<?php

// consume.php
// GUSER=yourAccountName@gmail.com GPASS="yourGmailPassword" php consume.php

require_once __DIR__.'/vendor/autoload.php';

$transport = new Swift_SpoolTransport(new \Swift_QueueSpool(
    (new \Enqueue\Fs\FsConnectionFactory('file://'.__DIR__.'/queue'))->createContext()
));

/** @var \Swift_QueueSpool $spool */
$spool = $transport->getSpool();
$spool->setTimeLimit(3);

$realTransport = (new Swift_SmtpTransport('smtp.gmail.com', 465, 'ssl'))
    ->setUsername(getenv('GUSER'))
    ->setPassword(getenv('GPASS'))
;

$spool->flushQueue($realTransport);
```

TODO:

* [ ] tests
* [ ] docs.
* [ ] TBD PsrProcessor

### CODE REVIEW

**Comment by makasim** [comment]
@fabpot Will you merge this? If so, I am going to add some tests and docs

File: lib/classes/Swift/QueueSpool.php [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+    /**
+     * @param PsrContext      $context
+     * @param PsrQueue|string $queue
+     */
+    public function __construct(PsrContext $context, $queue = 'swiftmailer_spool')
+    {
+        $this->context = $context;
+
+        if (false == $queue instanceof PsrQueue) {
```
Comment (Nyholm) [inline lib/classes/Swift/QueueSpool.php]: ===

File: lib/classes/Swift/QueueSpool.php [hunk 2]
```diff
@@ -0,0 +1,108 @@
+<?php
+
+use Interop\Queue\ExceptionInterface as PsrException;
+use Interop\Queue\PsrContext;
+use Interop\Queue\PsrQueue;
+
```
Comment (Nyholm) [inline lib/classes/Swift/QueueSpool.php]: Some class docs would be preferred.

File: lib/classes/Swift/QueueSpool.php [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+     */
+    private $queue;
+
+    /**
+     * @param PsrContext      $context
+     * @param PsrQueue|string $queue
+     */
+    public function __construct(PsrContext $context, $queue = 'swiftmailer_spool')
+    {
+        $this->context = $context;
```
Comment (Nyholm) [inline lib/classes/Swift/QueueSpool.php]: This could be moved to the bottom of the block. It will make the code a (tiny) bit easier to read.

**Comment by makasim** [comment]
@Nyholm done

**Comment by makasim** [comment]
Added amqp interop optimized spool mailer

**Pull request closed without merging**